### ISSUE TITLE
Add Time.diff, Date.add and DateTime.diff

### ISSUE THREAD

**Issue by josevalim** [elixir-lang/elixir#6183]
Note `NaiveDate.add/diff` and `Date.diff` already exists. We should also improve the documentation for using those functions as references and inline common use cases (unix and gregorian epoch).

**Comment by fxn** [comment]
@josevalim I think Time.diff has the same potential misuses of Time.add. The difference is ill-defined unless all you want is modulo arithmetic.

Date.add could have similar misuses, because sometimes addid 24h (in seconds, say), changes the date, but not always.

It's tempting to define them, but then you have to document they are naive in that sense. But then you are at square one. Which real world use case could serve as documentation example for instance?

**Comment by lau** [comment]
`Time` is "naive" just like `NaiveDateTime` in the sense that it does not have any time zone. Maybe that should be made more clear in the documentation.

**Comment by lau** [comment]
> Date.add could have similar misuses, because sometimes adding 24h (in seconds, say), changes the date, but not always.

`Date` is simple because there are no seconds, just whole days.

**Comment by fxn** [comment]
Yes, they are naive and that is known. But adding an API without a use case and with risk of being misused... is risky, can bite back later.

Not saying there are no use cases eh? Only I can't think of one, but can think of others in which people are going to have bugs.

I was thinking about chronographs, lap times and that kind of thing. There adding and substracting make sense, but that would need >24h in the general case.

Dates have whole days, yes, if units were days it would be well-defined, but the units I see defined are seconds and smaller. Is the plan to have date arithmetic only for day units?

**Comment by Qqwy** [comment]
I see no problem with Date. add, because it should only work with units of day-size and larger, which will make the results predictable. 

Time.diff has a problem however, which is what to do when the difference is 'more than twelve hours'. Note the apostrophes, as the difference between, say, 8:00 and 23:00 could be said to be 14, or 9, or even -9. I am not sure which version would be the best. Maybe we can  come up with a more descriptive way to call this so the taken approach is clear from the name.

@fxn: For laptimes, a different datastructure alltogether is needed. I think a Duration à la Joda-time would be wonderful, but something to either put in a library rather than the standard library, or to consider only after the current date/time stuff has settled somewhat (because od the ease of adding and difficulty of removing functionality)

**Comment by lau** [comment]
> Yes, they are naive and that is known. But adding an API without a use case and with risk of being misused... is risky, can bite back later.

Yes, I agree that that can be risky.

Thinking of `Time` as a simple, but very accurate, 24 hour wrist watch without complications (in the watch sense ie. no calendar, no stop watch, no moon phase) can make sense. You can set it forward or backwards by 30 seconds and there is no DST involved in that.

> Dates have whole days, yes, if units were days it would be well-defined, but the units I see defined are seconds and smaller. Is the plan to have date arithmetic only for day units?

I assume so. For smaller units it wouldn't make sense. Also in master there is `Date.diff` which uses an integer for whole days.

**Comment by fxn** [comment]
The Elixir community could be different here. If everyone is aware that Time is naive (which is very clear), and that in consequence the API works as you would expect in a naive type... then maybe misuses are going to be less likely, and people are going to switch to proper modules if the need arises.

I don't have context there, you guys know better for sure.

Agreed that day units for Date are fine. Also greater if edge cases are documented (eg, what means to add one year to a 29th of February, you know).

**Comment by lau** [comment]
> It's tempting to define them, but then you have to document they are naive in that sense. But then you are at square one. Which real world use case could serve as documentation example for instance?

I think the answer to the use case is closely related to most use cases of the `Time` type. Which is very much like the `TIME` type in SQL databases.

If people say when they usually wake up in the morning on weekdays, then `Time` could be appropriate. It is not on a particular day, it is for multiple days, so having a date in the type does not make sense. If you have two people that respond: one person responds `06:00` and another at `08:00`. Now let us say who wakes up first in the morning? We could just say that this is impossible because we don't know the time zone and call it a day. Or maybe we are OK with the results we can compute without having that data. Maybe all of the data is from people living in the same country and time zone. Maybe they are from all over the world, but we just want to compare when people wake up relative to their own time zone.

So who wakes up first (according to their own clock, and how they report it)? For that we could use a `diff` function. And once we have a `diff` function with its then a lot of the questions about `add` can be answered by being consistent with what is decided for `diff`.

Does it make sense to say that `8:00` is after `6:00`? I think so. Does it make sense to say that the difference is two hours or 7200 seconds? Again I think it does.

What if there was a DST changing event between 6:00 and 8:00? Just like with the functions in `NaiveDateTime` we are not concerned about that.

**Comment by fxn** [comment]
@lau forgot to mention, yeah the watch example makes sense. Could be a real watch, maybe the watch of a game in which DST does not even exist.

**Comment by fxn** [comment]
@lau I am in Kiev, in a conf, writing with my phone in a flaky hotel wifi :). We're crossing messages, but the discussion looks productive to me!

**Comment by josevalim** [comment]
Some clarifications:

  * The second argument for `Date.add` is going to be an integer representing days. There should be no ambiguity over its use.
  * For `Time.diff`, if the left side is smaller than the right side, a negative number will be returned.

**Comment by michalmuskala** [comment]
I'm a bit late to the discussion.

Considering we're planning to support using `diff` to replace `to_gregorian`, etc would it be a good idea for `diff` to accept atoms in the second argument for well-known points in time? This way I could call `Date.diff(some_date, :gregorian_epoch)` or `Date.diff(some_date, :unix_epoch)`. This should also make it easier to use and faster than passing the full date structs for those points in time.

**Comment by josevalim** [comment]
@michalmuskala I like this idea but I am worried about how we will have to support those values across the board. Should it be only for diff? What about add? Since those two functions actually form the proper conversion pairs? Maybe we should have Time.instant, NaiveDateTime.instant for capturing certain named instants but I am not sure (yet).

**Comment by michalmuskala** [comment]
Yes, I think we would need to support it both in `diff` and `add`. One problem with having a separate `instant/1` is that we change the initial API we wanted:
```elixir
Date.to_gregorian_days(date)
```
into
```elixir
Date.diff(date, Date.instant(:gregorian_epoch))
```
It might be a little difficult to "find" this combination of `diff` and `instant`.

**Comment by fxn** [comment]
I have been think about this since yesterday and [asked in Twitter about use cases for time arithmetic](https://twitter.com/fxn/status/871256720256925697), with no compelling reponse so far.

Let me summarize my thoughts regarding `Time`:

1. I totally see use cases for _representing_ time that way. Some could be risky, because in my country for example there is one night in which 02:30:00 just does not exist. If interpreted as _time_, it might not valid. But... I can see some time tables being implemented with time, and I can see a view template that only shows time to receive a time from its caller (who is responsible for its correctness, maybe coming from a DST-aware source).

2. I cannot imagine real, practical uses cases for doing _any_ arithmetic with them. In general with APIs I prefer to have a clear use case. Adding for the sake of adding is not good in my experience, and as we know removing is harder. Also, the _mere_ existence of the function may drive misusages. So I am really dubious there.

3. You can always add a function if someone opens a PR with a real use case.

4. I can think of use cases when the time means duration: how long did this took? Lap times, song duration, etc. There, arithmetic makes sense. But that doesn't work well with the 24h constraint, 36:12:27 should be valid.

That said, you guys have a tone more context about these modules, their history, where do they go, and the expectations and usage patterns of Elixir programmers. So whatever you end up implementing is going to be good 😄 .

**Comment by josevalim** [comment]
> I can think of use cases when the time means duration: how long did this took? Lap times, song duration, etc. There, arithmetic makes sense. But that doesn't work well with the 24h constraint, 36:12:27 should be valid.

This is a very good example! Unfortunately `Time` in Elixir is meant to represent instants, according to the [ISO 8601 spec](https://en.wikipedia.org/wiki/ISO_8601) and therefore does not fit the use case above. It sounds the example above is like a duration according to ISO 8601, which we do not (yet) support in Elixir.

And even though we consider `Time` to be naive, which rules out item 1, we won't add `Time.add/3` for now per item 2. It can always be added later per item 3.

**Comment by fxn** [comment]
But `Time.diff` is also arithmetic.

**Comment by josevalim** [comment]
We need to provide a reference though for computing the amount of seconds between two time instants. I can see how it is easy to run into issues with `Time.add/3` due to overlaps but those are simpler to document in `Time.diff/3`.

Alternatively we can investigate what libraries such [as JodaTime does](http://www.joda.org/joda-time/apidocs/org/joda/time/LocalTime.html).  We can see they assume UTC for all times. It is not documented what overlaps does though.

**Comment by lau** [comment]
@fxn The `Time` type in Elixir is equivalent to the `TIME` type (without "timezone") in Postgres.

I provided an example of calculations based on the time people wake up and comparing the two. https://github.com/elixir-lang/elixir/issues/6183#issuecomment-305996683

Another more generic example:

* You have a time field in a postgres database and want to compare two fields. How will you do that without a diff function?

If you have a postgres database with a `time` field is it not a real use case to compare the two?

> I can think of use cases when the time means duration

A duration type would be interesting but that's not what the `Time` type is for. Think of time as a "generic" time of day that can be applied to any date. Or as I wrote yesterday: a watch. Another typical example is a time for an alarm clock. The clock has no concept of a date or a time zone. The time it is set to go off is just a time of day.

> I totally see use cases for representing time that way. Some could be risky, because in my country for example there is one night in which 02:30:00 just does not exist.

`NaiveDateTime` already has:
- an `add` function
- a `diff` function
- No concept of time zones

So you could use `NaiveDateTime` to do take 1:59 on that day, add some seconds to it and get 2:30. This is already something that exists in the library today.

Some times you do not know the time zone but want to calculate something anyway. We already have functions for that in `NaiveDateTime`.

**Comment by lau** [comment]
> Date.diff(some_date, :unix_epoch)

That might not be a good idea. Unix epoch is on a specific time of day, specifically in UTC, but `Date` has no time zone.

`~D[0000-01-01]` is more simple than `Date.instant(:gregorian_epoch)` is it not?

For internal logic using the gregorian epoch wouldn't it work to do a module attribute like
`@gregorian_epoch_date ~D[0000-01-01]`?

**Comment by lau** [comment]
> Date.diff(some_date, :unix_epoch)

Another approach if you want to compare a date to the date of the Unix epoch:

To get unix epoch in Elixir today you can do: `DateTime.from_unix!(0)`

In `master` this should give you the difference in whole days:
`Date.diff(DateTime.from_unix!(0), ~D[2017-06-04])`

That would just use the date part of the `DateTime` so `1970-01-01`.

**Comment by michalmuskala** [comment]
As far as I understand "downcasting" on comparisons, etc was deprecated.

**Comment by lau** [comment]
> As far as I understand "downcasting" on comparisons, etc was deprecated.

Why? When did that happen?

**Comment by michalmuskala** [comment]
This commit: https://github.com/elixir-lang/elixir/commit/d017c5ab1222fef2f114f327767c33c9310f7b52
The reason is probably somewhere in the related PR discussion: https://github.com/elixir-lang/elixir/pull/5841

**Comment by josevalim** [comment]
> ~D[0000-01-01] is more simple than Date.instant(:gregorian_epoch) is it not?

Yes, I would say so. Hence why I am holding back on the instant feature. We can discuss the pros and cons, but definitely not part of this release.

> As far as I understand "downcasting" on comparisons, etc was deprecated.

That's incorrect. Downcasting was deprecated only on to_erl and to_iso because of calendar ambiguity. `day_of_the_week`, `leap_year?`, `days_in_month`, `to_string`, `compare`, etc all do the downcasting still.

**Comment by Qqwy** [comment]
A quick remark: The gregorian epoch was on `~D[0001-01-01]`, the 'year zero' that is included in ISO8604 actually points to 1 BC. The whole concept of '0' postdates the invention of the julian and gregorian calendars. 

This is exactly the kind of edge cases that happen when considering calendars. Not sure what to conclude from it, but I wanted to note it either way.
